### PR TITLE
Implement torch.nextafter

### PR DESCRIPTION
**Pull request by muthuArivoli**
Related to #38349.

### CODE REVIEW

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -4556,6 +4556,28 @@ def merge_dicts(*dicts):
     tensor([-0.0090,  0.2262,  0.0682,  0.2866, -0.3940])
 """.format(**common_args))
 
+add_docstr(torch.nextafter,
+           r"""
+nextafter(input, other, *, out=None) -> Tensor
+
+Return the next floating-point value after input towards other, elementwise.
```
Comment (mruberry) [inline torch/_torch_docs.py:4563]: :attr:\`input\` and :attr:\`other\` here and in the following paragraph.

File: test/test_torch.py [hunk 2]
```diff
@@ -16987,6 +16987,23 @@ def test_lcm(self, device, dtype):
         expected = np.lcm(a.cpu().numpy(), b.cpu().numpy())
         self.assertEqual(actual, expected)
 
+    @onlyOnCPUAndCUDA
+    @dtypes(torch.float32, torch.float64)
```
Comment (mruberry) [inline test/test_torch.py:16991]: `@dtypesIfCUDA(torch.float16, torch.float32, torch.float64)` can be added to test your half implementation on CUDA only

File: test/test_torch.py [hunk 3]
```diff
@@ -16987,6 +16987,23 @@ def test_lcm(self, device, dtype):
...
+    @unittest.skipIf(not TEST_NUMPY, "NumPy not found")
+    def test_nextafter(self, device, dtype):
+        # Test special cases
+        t1 = torch.tensor([0, 0, 10], device=device, dtype=dtype)
+        t2 = torch.tensor([inf, -inf, 10], device=device, dtype=dtype)
```
Comment (mruberry) [inline test/test_torch.py:16996]: Better add nan, too

File: test/test_torch.py [hunk 4]
```diff
@@ -16987,6 +16987,23 @@ def test_lcm(self, device, dtype):
...
+        actual = torch.nextafter(t1, t2)
+        expected = np.nextafter(t1.cpu().numpy(), t2.cpu().numpy())
```
Comment (mruberry) [inline test/test_torch.py:16998]: Better test the flip of these `torch.nextafter(t2, t1)` as well

**Review by mruberry — COMMENTED** [review]
This looks really good, @muthuArivoli, nice work! I just made a few suggestions for the test and doc.

**Comment by muthuArivoli** [comment]
@mruberry, I removed the half implementation since it didn't work. I also have made all the other changes requested.

**Comment by mruberry** [comment]
Everything looks very good - would just add 'nan' as another special value in the test (along with inf/-inf)?

**Comment by muthuArivoli** [comment]
> Everything looks very good - would just add 'nan' as another special value in the test (along with inf/-inf)?

I added nan in the test after that test mentioned. Other tests seemed to use `assertTrue` with `isnan` to test for nan, since I think `nan!=nan`, which is why I used it.

**Comment by mruberry** [comment]
> > Everything looks very good - would just add 'nan' as another special value in the test (along with inf/-inf)?
> 
> I added nan in the test after that test mentioned. Other tests seemed to use `assertTrue` with `isnan` to test for nan, since I think `nan!=nan`, which is why I used it.

Aha! Thank you for pointing that out. My mistake.

**Review by mruberry — APPROVED** [review]
Awesome work, @muthuArivoli! Thanks for this PR.

I just updated https://github.com/pytorch/pytorch/issues/38349, too, if you're interested in other functions.

**Comment by mruberry** [comment]
We have some internal mobile builds that are failing after this PR:

```
aten/src/ATen/cpu/vec256/vec256_base.h:389:16: error: no member named 'nextafter' in namespace 'std'; did you mean simply 'nextafter'?
      ret[i] = std::nextafter(values[i], b[i]);

ten/src/ATen/native/cpu/BinaryOpsKernel.cpp:749:22: error: no member named 'nextafter' in 'at::vec256::(anonymous namespace)::Vec256<float>'
            return a.nextafter(b);
```

cc @dreiss 

I'm not very familiar with mobile but I can follow-up with them to get a better understanding of what's going on here and how to address this issue. One option, if you like, would be to removed the CPU vectorized code paths for now, add this function, and then add the vectorized code paths back once we have a plan to deal with this mobile issue. Either that or we can wait for the mobile team to weigh-in.

**Comment by mruberry** [comment]
Now that https://github.com/pytorch/pytorch/pull/42291 is in, this PR will need analogous updates.

**Comment by muthuArivoli** [comment]
Just updated with the mobile fixes. I realized however, that I haven't done anything for autograd. Do I need to do anything for autograd, and if so, how would I, since I don't think the gradient is well defined for nextafter?

**Comment by mruberry** [comment]
> Just updated with the mobile fixes. I realized however, that I haven't done anything for autograd. Do I need to do anything for autograd, and if so, how would I, since I don't think the gradient is well defined for nextafter?

You could add a stub like this:

https://github.com/muthuArivoli/pytorch/blob/ab5c39810aa4d914b1f4e97f2873976cd3415b56/tools/autograd/derivatives.yaml#L220

Which would throw a nicer error message if someone tries to backward through nextafter.

**Comment by muthuArivoli** [comment]
> > Just updated with the mobile fixes. I realized however, that I haven't done anything for autograd. Do I need to do anything for autograd, and if so, how would I, since I don't think the gradient is well defined for nextafter?
> 
> You could add a stub like this:
> 
> https://github.com/muthuArivoli/pytorch/blob/ab5c39810aa4d914b1f4e97f2873976cd3415b56/tools/autograd/derivatives.yaml#L220
> 
> Which would throw a nicer error message if someone tries to backward through nextafter.

Thanks! Just added that.

**Comment by muthuArivoli** [comment]
@mruberry Could you run this against the internal mobile builds?

**Comment by mruberry** [comment]
Android builds look good. Starting the land process.
